pkg/trojan/message_test.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2020 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package trojan
+
+import (
+	"bytes"
+	"encoding/binary"
+	"math/big"
+	"reflect"
+	"testing"
+
+	chunktesting "github.com/ethersphere/bee/pkg/storage/testing"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// arbitrary targets for tests
+var t1 = Target([]byte{57, 120})
+var t2 = Target([]byte{209, 156})
+var t3 = Target([]byte{156, 38})
+var t4 = Target([]byte{89, 19})
+var t5 = Target([]byte{22, 129})
+var testTargets = Targets([]Target{t1, t2, t3, t4, t5})
+
+// arbitrary topic for tests
+var testTopic = NewTopic("foo")
+
+// newTestMessage creates an arbitrary Message for tests
+func newTestMessage(t *testing.T) Message {
+	payload := []byte("foopayload")
+	m, err := NewMessage(testTopic, payload)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return m
+}
+
+// TestNewMessage tests the correct and incorrect creation of a Message struct
+func TestNewMessage(t *testing.T) {
+	smallPayload := make([]byte, 32)
+	m, err := NewMessage(testTopic, smallPayload)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// verify topic
+	if m.Topic != testTopic {
```

Comment:
I would have expected a separate test for topic. Or is that implicitly covered by `padBytes` below?

Reply:
unit tests for topic could be added. right now it's covered by the tests that implicitly use topic functions, but we could have separate unit tests anyway

Reply:
please consider doing topic sealing as per the spec instead of testing temporary stuff

Reply:
it seems to me that "sealing" belongs in the pss package? wouldnt putting it here be mixing areas of responsibility?

Reply:
So I read the sealing part in the book. As far as I understand now the sealing only has to do with topic obfuscation. I presume the effect of that is that for every message received you have to iterate over all registered topic and hash them together with the payload to find one that matches.

That begs the question; should it be possible to utilize the trojan functionality without being forced to spend cycles on this. And in extension of that, should topic obfuscation be an optional feature - if I don't care about being opaque, should I be forced to be.

Reply:
What does "sealing" mean BTW?

Reply:
`sealedtopic` on the wire is `hash(topic|payload)` so that the receiver would need for every registered `topic` try `hash(topic|payload)` to see if one of them matches `sealedtopic` ... if i understand correctly.